mlflow/gateway/providers/openai.py

```diff
@@ -121,18 +133,214 @@ async def chat_stream(
+                payload=self._add_model_to_payload_if_necessary(
+                    {
+                        **payload,
+                        "messages": messages,
+                    }
+                ),
+            )
+
+        elif any(t["type"] == "uc_function" for t in payload.get("tools", [])):
+            updated_tools = []
+            uc_func_mapping = {}
+            for tool in payload.get("tools", []):
+                if tool["type"] == "uc_function":
+                    function_name = tool["uc_function"]["name"]
+                    function = workspace_client.functions.get(function_name)
+                    param_metadata = get_func_schema(function)
+                    t = {
+                        "type": "function",
+                        "function": param_metadata,
+                    }
+                    uc_func_mapping[t["function"]["name"]] = function
+                    updated_tools.append(t)
+                    continue
+                else:
+                    updated_tools.append(tool)
+
+            payload["tools"] = updated_tools
+
+            messages = payload.pop("messages")
+            uc_func_calls = []
+            user_tool_calls = []
+            resp = None
+            should_break = False
+            for _ in range(20):
+                if should_break:
+                    if uc_func_calls:
+                        resp["choices"][0]["message"]["content"] = (
+                            user_msg + "\n\n" + join_uc_functions(uc_func_calls)
+                        )
+
+                    if user_tool_calls:
+                        resp["choices"][0]["message"]["tool_calls"] = user_tool_calls
+                    break
+
+                resp = await send_request(
+                    headers=self._request_headers,
+                    base_url=self._request_base_url,
+                    path="chat/completions",
+                    payload=self._add_model_to_payload_if_necessary(
+                        {
+                            **payload,
+                            "messages": messages,
+                        }
+                    ),
+                )
+                token_usage_accumulator.update(resp.get("usage", {}))
+                # TODO to support n > 1.
+                assistant_msg = resp["choices"][0]["message"]
+                tool_calls = assistant_msg.get("tool_calls")
+                if tool_calls is None:
+                    if uc_func_calls:
+                        resp["choices"][0]["message"]["content"] = prepend_uc_functions(
+                            resp["choices"][0]["message"]["content"], uc_func_calls
+                        )
+
+                    if user_tool_calls:
+                        resp["choices"][0]["message"]["tool_calls"] = user_tool_calls
+
+                    break
+
+                tool_messages = []
+                for tool_call in tool_calls:  # TODO: should run in parallel
```

Comment:
I don't think parallelism is a blocker for merging this PR or for MLflow 2.14.0, but can we file a ticket to add it?

Reply:
filed a ticket, will remove this comment.